pkg/cmd/pipelinerun/logs.go

```diff
@@ -0,0 +1,269 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pipelinerun
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"github.com/tektoncd/cli/pkg/cli"
+	"github.com/tektoncd/cli/pkg/logs"
+	v1alpha1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+	"github.com/tektoncd/pipeline/pkg/reconciler/v1alpha1/taskrun/resources"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	typedv1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+const (
+	msgPipelineNotFoundErr = "Error in retrieving Pipeline"
+	msgPRNotFoundErr       = "Error in retrieving PipelineRun"
+)
+
+type taskRunMap map[string]*v1alpha1.PipelineRunTaskRunStatus
+
+type taskRun struct {
+	name   string
+	status v1alpha1.TaskRunStatus
+}
+
+type PipelineRunLogs struct {
+	Clients   *cli.Clients
+	Out       io.Writer
+	Err       io.Writer
+	Streamer  func(typedv1.PodInterface) logs.PodLogStreamer
+	Name      string
+	NS        string
+	Tasks     []string
+	Follow    bool
+	InitSteps bool
+}
+
+func logCommand(p cli.Params) *cobra.Command {
+	var (
+		follow  bool
+		allLogs bool
+		tasks   []string
+		eg      = `
+  # show the logs of PipelineRun named "foo" from the namesspace "bar"
+  tkn pipelinerun logs foo -n bar
+
+  # show the logs of PipelineRun named "microservice-1" for task "build" only, from the namespace "bar"
+  tkn pr logs microservice-1 -t build -n bar
+   `
+	)
+
+	c := &cobra.Command{
+		Use:                   "logs",
+		DisableFlagsInUseLine: true,
+		Short:                 "Show the logs of PipelineRun",
+		Example:               eg,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			var prName string
+
+			if len(args) == 0 {
+				return errors.Errorf("Missing PipelineRun name parameter, please specify the pipeline name.\nCheck for help \n")
+			}
+			prName = args[0]
+
+			cs, err := p.Clients()
+			if err != nil {
+				return err
+			}
+
+			prLogs := &PipelineRunLogs{
+				Clients:   cs,
+				Out:       cmd.OutOrStdout(),
+				Err:       cmd.OutOrStderr(),
+				Streamer:  logs.DefaultStreamer,
+				Name:      prName,
+				NS:        p.Namespace(),
+				Tasks:     tasks,
+				Follow:    follow,
+				InitSteps: allLogs,
+			}
+
+			return prLogs.Fetch()
+		},
+	}
+
+	c.Flags().BoolVarP(&follow, "follow", "f", false, "stream the live pipeline run logs")
+	c.Flags().BoolVarP(&allLogs, "all", "a", false, "show all logs including init steps injected by tekton")
+	c.Flags().StringSliceVarP(&tasks, "only-task", "t", []string{}, "show the logs for mentioned task only")
+
+	return c
+}
+
+func (p *PipelineRunLogs) Fetch() error {
```

Comment:
I would make that a *standard* func that takes the struct described above as parameters, sthg like

```
func fetchLogs(cs *cli.Clients, opts logOptions, io ioStruct) error {
```

Reply:
Definitely :+1: :metal: